### PR TITLE
CMT merge

### PR DESCRIPTION
**Pull request by RonRahaman**
This is the first update to the CMT code since November.  It includes many improvements to particle tracking, among other things.  

This branch results from two merges and a rebase between the UFCCMT BitBucket repo and the Nek5000 GitHub repo.  Let these repos be called 'UFCCMT' and 'Nek5000' respectively. Our workflow was the following:
- The directories of UFCCMT was restrucuted using 'git filter-branch'.  This was necessary because UFCCMT had the directory structure of the SVN repo.  
- The first merge was performed:
  1. We identified the last time UFCCMT had been synced with Nek5000.  That sync was a manual copy-paste of source files, not a Git merge.  
  2. A new branch (UFCCMT:cmt-sync) was created on UFCCMT:master at the downstream commit identified in (1).  
  3. A new branch (Nek5000:cmt-merge-1) was created on Nek5000:develop at the downstream commit identified in (1).  
  4. We merged UFCCMT:cmt-sync into Nek5000:cmt-merge-1
- A rebase was performed:
  1. A new branch (UFCCMT:cmt-rebase-1) was created at the most recent commit of UFCCMT:master.
  2. UFCCMT:cmt-rebase-1 was rebased on top of Nek5000:cmt-merge-1
  3. The rebased UFCCMT:cmt-rebase-1 was pushed to Nek5000 as a new branch (Nek5000:cmt-rebase-1)
- A second and final rebase was performed:
  1. A new branch (Nek5000:cmt-merge-2) was created at the most recent commit of Nek5000:develop.
  2. Nek5000:cmt-rebase-1 was merged into Nek5000:cmt-merge-2

### CODE REVIEW

**Comment by RonRahaman** [comment]
Right now, the CMT example problems are in cmt-cases/.  They should probably be moved elsewhere, either in short_tests/ or in the NekExamples repo.  

Our intent is to have regression tests for cmt-cases/ at some point in the near future.  Thoughts?

cc: @fischer1 @hackljf @stgeke

**Comment by stgeke** [comment]
I suggest the following changes:
- move cmt out of core into ext/cmt (create an ext top-level directory)
- move cmt_cases into examples and/or short-tests 
- use a pre-processor flag IFCMT instead of logical flag in makenek
- remove all cmt specific code from core (we may need to do some changes to support extensions like this)

**Comment by stgeke** [comment]
Why do we want to include something like CMT (which is experimental feature) into our stable release branch?

**Comment by pschlatt1** [comment]
I would agree on moving CMT out of core; but perhaps in the 3rd-party directory?

**Comment by RonRahaman** [comment]
Those are great ideas, but moving cmt/ out of core/ should be a separate PR.  This PR contains bugfixes to the existing cmt/ code.  On the other hand, moving cmt/ is an enchancement.  PRs should represent atomic concepts.

**Comment by stgeke** [comment]
Is this PR atomic? It looks more like an initial merge to me.

**Comment by stgeke** [comment]
Can somebody remind me again about the intention of having a top-level 3rd-party directory?

**Comment by pschlatt1** [comment]
my understanding is exactly for module like cmt; non-essential add-ons predominantly developed, maintained and used by non-core developers.

**Comment by stgeke** [comment]
Really? I don't think we should host external source code in our project.

**Comment by pschlatt1** [comment]
I guess it depends on the definition of external. We were thinking of e.g. Arnoldi which is non-essential, but still widely used and should be in-sync with the main development. But I might be wrong, of course.

**Comment by RonRahaman** [comment]
It's not external.  @fischer1 is a co-PI of the CCMT at University of Florida.

**Comment by stgeke** [comment]
Don't get me wrong - I'm not saying that CMT should be a separate project. I just thought it's a good opportunity to bring the discussion about extensions, plug-ins, toolboxes up again. This is something we need to sort out pretty soon.

**Comment by hackljf** [comment]
Dear Ron, Stefan, Prof. Schlatter, and others,

I really appreciate your hard work getting CMT-nek integrated into nek5000
cleanly and consistently. I am recovering from surgery and busy with our DG
solver for Navier-Stokes and artificial viscosity as well as bringing some
old examples back up to date.

CMT-nek in its current form was designed as an "execution branch" from an
otherwise conventional run of nek5000. Paul and my former coworker, Mrugesh
Shringarpure, worked out the current method. The only changes to core/ were
meant to be IFCMT in the advance loop and nekuic. Unfortunately, restarting
required the odd situation where we needed the T field, but couldn't run
with the dssums done IFHEAT=T. Paul committed some workarounds to that. I
apologize that it got a little messier. I defer to whatever design
decisions you guys make. Just please teach us in Florida how to run with
them.

I will get you at least one example for short_cases today. If you change
makenek, please teach me what to do.

## 

Jason Hackl

**Comment by RonRahaman** [comment]
@stgeke, I agree that the discussion about 3rd party modules is a good one to have.  

I was under the impression that the CMT improvements need to make their way into Nek5000 one way or another, and that's my only concern.  

I will of course defer to your judgement regarding:
- Whether the CMT bugfixes should be included in v16.0.0
- Whether CMT should live in 3rd_party/ eventually

Once you've decided what to do, I'd appreciate it if you raised an Issue outlining exactly what we need to do to get the CMT changes back into Nek.

Thanks!

**Comment by stgeke** [comment]
I think this PR should be merged into `develop` and CMT (which is in an experimental stage) should be deleted from the `release` branch, at least for now.

**Comment by hackljf** [comment]
I wouldn't worry about CMT bug fixes in 16.0.0. As long as the world at
large can download the homentropic vortex case from short_cases and run it
for at least a little while, that's fine. I am OK with removing CMT from
release and leaving it in develop AS LONG AS the develop branch is publicly
available to students at UF and I can issue pull requests to it.

I'm not sure what the intended spirit of 3rd_party is. It would be nice for
Paul to decide whether or not CMT-nek lives there.

## 

Jason Hackl

On Mon, Sep 5, 2016 at 10:50 AM, Stefan Kerkemeier <notifications@github.com

**Comment by RonRahaman** [comment]
Yes, @stgeke, I respect your opinion and you have some very valid points, but the decision to remove CMT should not be taken lightly.  @fischer1 had originally put CMT into Nek, so it should be his decision.  

IMO, it doesn't make sense to remove CMT from 'release/v16.0.0' just to add it back into 'develop' post-release.

**Comment by stgeke** [comment]
Yes, I am not the one who decides. Keep in mind: In the past we had a development but no release branch. That's why we have some experimental things (not just cmt) in the release today. IMO, something we need to clean-up.

**Comment by RonRahaman** [comment]
If you'd like to start removing features, I'd appreciate it if you stated your intentions in detail on the [Issues page](https://github.com/Nek5000/Nek5000/issues).  If you have a clear plan, I'm happy to support you.

**Comment by RonRahaman** [comment]
> I suggest the following changes:
> 
> move cmt out of core into ext/cmt (create an ext top-level directory)
> move cmt_cases into examples and/or short-tests
> use a pre-processor flag IFCMT instead of logical flag in makenek
> remove all cmt specific code from core (we may need to do some changes to support extensions like this)

Right now, I'm happy to do the following.  They're unequivocably good ideas:
- [ ]  Move all cmt_cases into NekExamples
- [ ]  Add the CMT verificaiton to short_cases, Travis, and buildbot.  @hackljf just shared these with me
- [ ]  Replace IFCMT in makenek with a preprocessor flag in makenek.inc

I'm only comfortable doing the other action items (esp. deleting this feature from 'release/v16.0.0') after input from @fischer1.  @stgeke and @hackljf, is this satisfactory?

**Comment by RonRahaman** [comment]
@stgeke, do you have a preference for the preprocessor symbol we use in PPLIST?  The symbols "CMT" and "IFCMT" are already in use.

**Comment by fischer1** [comment]
I envision one or more compressible variants, perhaps to be merged into a single option down the road.   Compressible is something for which there is significant demand and will ultimately be on an equal footing with, say, low Mach as an option.  

I very much like the idea of supporting Arnoldi as well.  The one difficulty I have with adjoints and Arnoldi is that there seems to be a large case-to-case dependency (correct me if I'm wrong), which makes it difficult to live in the source, since one needs to modify things from one problem to the next.  (At least, that was my impression with adjoints --- and I merely hazard a guess that the same carries over for Arnoldi.)

My inclination is that, if it has broad use, is highly repeated code that would otherwise be duplicated in 100s of .usr files, then it probably belongs in "core".   The idea of the CMT subdirectory was simply to give the CMT group room to work, with a lightweight connection to the main source.   It's important that their code be in the Nek5000 repo.  I don't know that we need to put it in 3rd party.   3rd party, in my view, would be for something that is really distinct from Nek (e.g., moab, lapack) as opposed to something that is closely integrated with Nek.

I _think_ Arnoldi would be in the spirit of many of the post-processing / ancillary routines that are in postpro and navier5.f.  If it has a relatively stable interface, then it should be in core.

**Comment by fischer1** [comment]
@hackljf  Jason - can you resolve the conflicts for this merge request?  Paul

**Comment by stgeke** [comment]
I think of CMT as an extension which should be encapsulated with a well defined interface. To accept the pull request I recommend the following changes:

0) Pull in the latest changes and resolve the conflicts
1) Eliminate IFCMT in makenek and add a preprocessor symbol (CMT) instead
2) Compile source code in core/CMT if CMT is defined
3) Use the existing t-array for fields like (phig, …), no need to define extra arrays for them
4) Remove all CMT stuff from core except the top-level driver call in nek_advance() and ifcmt in INPUT
    - With (3) you don’t need to modify nekuic() anymore
    - Instead of putting your data into NEKUSE have your own CMTUSE which you include in the .usr file
    - Not sure how you could avoid the changes in nekasgn(), but may be you have a good idea?

We can consider this pull request for v16.0.0 if you can implement the suggested changes within the next 14 days.

**Comment by hackljf** [comment]
Thank you so much for your management and patient suggestions. With Ron's
or somebody else's help introducing a preprocessor symbol for step 1, I
will have all of these done in 14 days. I think I can also write a
dedicated nekasgn for CMTUSE and leave it somewhere in the cmt/ directory's
source code.

What is the maximum value of ldimt I can use?

## 

Jason Hackl

**Comment by stgeke** [comment]
I think of CMT as an extension which should be encapsulated with a well defined interface. To accept the pull request I recommend the following changes:

0) Pull in the latest changes and resolve the conflicts
1) Eliminate IFCMT in makenek and add a preprocessor symbol (CMT) instead
2) Compile source code in core/CMT if CMT is defined
3) Use the existing t-array for fields like (phig, …), no need to define extra arrays for them
4) Remove all CMT stuff from core except the top-level driver call in nek_advance() and ifcmt in INPUT
	- With (3) you don’t need to modify nekuic() anymore
	- Instead of putting your data into NEKUSE have your own CMTUSE which you include in the .usr file
	- Not sure how you could avoid the changes in nekasgn(), but may be you have a good idea?

Any chance to implement the suggest changes within the next 2-3 weeks? It would be good to include them in v16.0.0.

**Comment by stgeke** [comment]
Forgot to add: You should also eliminate the CMT stuff in SIZE. I don't think this is needed anymore with the preprocessor flag and the t-array.

It's up to you (the user) to set a value for ldimt in SIZE.

**Comment by stgeke** [comment]
I can help you with the preprocessor stuff and the build system. Just go ahead and implement the suggest changes and I'll do the rest.

**Comment by hackljf** [comment]
right, but I'm asking if, in classic nek, ldimt is a number that is
typically less than 7 or not. We need at least this many variables, and
there's a good chance we'll need more.

**Comment by hackljf** [comment]
thank you for your help.

**Comment by stgeke** [comment]
The limit is and was always the available memory.

**Comment by RonRahaman** [comment]
This branch is back in sync with the upstream changes and passes all the tests.  Can we merge it now and work on the IFDEFs as a separate Issue?  That way, someone else can claim the issue and decouple it from these changes.  

cc: @stgeke @hackljf @fischer1

**Comment by hackljf** [comment]
That gets my vote.

**Comment by stgeke** [comment]
This PR contains precore and postcore in tools. What's that?

**Comment by stgeke** [comment]
You did a good job!

Please do the following minor corrections:
- prefix the routines you add to the .usr file (e.g. usrflt -> cmt_usrflt) to avoid name conflicts
- remove core/README
- remove tools/precore
- remove tools/postcore
- remove short_tests/CMT/inviscid_vortex/fort.144
- use lastest SIZE.template for short_tests/CMT/SIZE

**Comment by RonRahaman** [comment]
@hackljf, I'm making some fixes so that PPLIST will work in NekTests as intended.  

Some necessary changes to short_tests/lib were just merged in #158.  They introduced some expected conflicts in this PR that I will resolve.  

I also need to make a few changes to the CmtInviscidVortex class in NekTests.py.  Stay tuned for a PR in ufccmt/Nek5000

**Comment by RonRahaman** [comment]
To update everyone, NekTests.py has been updated to support the new compilation options for CMTNEK.  

Previously, NekTests would set IFCMT=true in makenek to run the CmtInviscidVortex test case.  Now, NekTests will set CMTNEK in PPLIST to run that test case.  

These changes were merged into UFCCMT/Nek5000:cmt-merge-2 and automagically updated in this PR.

**Comment by stgeke** [comment]
The cmt/README.md looks pretty specific to me. Is this file really needed?
Is there anything else we need to do before we can merge this PR?

**Comment by hackljf** [comment]
Removed and merged

**Comment by stgeke** [comment]
@fischer1 I recommend to merge this PR after the conflict was resolved.

**Comment by stgeke** [comment]
@hackljf Please resolve the conflict.

**Comment by hackljf** [comment]
I don't understand conflicts yet at all. When I tried to emulate some of
the changes in v16.0.0.0's ic.f that were not in cmt-merge-2's ic.f. it
broke PnPn (but everything else passed UFCCMT's regression testing).

Whatever the version of ic.f I am conflicting with looks like (and I can't
tell if it's the one I see in release/v16.0.0.0 or not), I have fulifilled
my promise of removing ALL cmt code and declarations from it with the
UNIQUE exception of

 #ifdef CMTNEK
         ifgett=.true. ! CMT-nek still not compatible with IFHEAT
#endif

near the end of sioflag. Otherwise, the characters "c," "m" and "t" do not
appear near each other because of CMT-nek.

 I am deeply sorry I don't know what else I can do.

-- 

Jason Hackl

File: core/ic.f [hunk 1]
```diff
@@ -56,14 +50,6 @@ subroutine setics
 
       jp = 0                  ! Set counter for perturbation analysis
 
-      if (ifcmt) then
```
Comment (RonRahaman) [inline core/ic.f:59]: The only conflicts between this branch (cmt-merge-2) and release/v16.0.0 occurs here.  The diff looks like this, where the upper referes to the changes in cmt-merge-2 and the lower refers to release/v16.0.0:

```diff
++<<<<<<< cmt-merge-2
 +      irst = param(46)        ! for lee's restart (rarely used)
 +      if (irst.gt.0)  call setup_convect(2)
 +
++=======
+       if (ifcmt) then
+          call rzero(phig,ltott)
+          call rzero(csound,ltott)
+          call rzero(vtrans,ltott*ldimt1)
+          call rzero(vdiff ,ltott*ldimt1)
+          call rzero(u,ntotcv)
+       endif
++>>>>>>> release/v16.0.0
```

**Review by RonRahaman — COMMENTED** [review]
@hackljf, you resolve the merge on the command line like this.  Suppose you have the following remotes defined:
* 'origin' points to https://github.com/hackljf/Nek5000.git 
* 'upstream' points to https://github.com/Nek5000/Nek5000.git 
* 'ufccmt' points to https://github.com/UFCCMT/Nek5000.git.  

Then you do the following:
```
git checkout cmt-merge-2
git pull ufccmt cmt-merge-2    # Make sure your local branch is up-to-date
git pull upstream release/v16.0.0    # Uh-oh!  Conflicts!
```

As you noted, the conflicts are in core/ic.f.  The conflicts are limited to the lines noted here (in the conversation for this review).  Luckily, these are small conflicts and you can just clean it up in your text editor (no need to use a fancy merge tool).  Once you've cleaned-up the file to your liking, you do the following

```
git add core/ic.f    # Mark the conflicts as resolved
git commit    # Commit the merge
git push origin cmt-merge-2    # Push your changes to your fork
```
Then you submit following PR:
https://github.com/UFCCMT/Nek5000/compare/cmt-merge-2...hackljf:cmt-merge-2

Once the above PR is merged, then this PR (the one we're discussing) will be updated and shouldn't have any conflicts.

**Comment by hackljf** [comment]
Thank you. I will tend to this and follow your instructions in the next
three hours

**Pull request merged**